gson/src/test/java/com/google/gson/GsonBuilderTest.java

```diff
@@ -254,4 +255,38 @@ public void testSetStrictness() throws IOException {
     assertThat(gson.newJsonReader(new StringReader("{}")).getStrictness()).isEqualTo(STRICTNESS);
     assertThat(gson.newJsonWriter(new StringWriter()).getStrictness()).isEqualTo(STRICTNESS);
   }
+
+  @Test
+  public void testRegisterTypeAdapterForObjectAndJsonElements() {
+    final String ERROR_MESSAGE = "Cannot override built-in adapter for ";
+    Type[] types = {
+            Object.class,
+            JsonElement.class,
+            JsonArray.class,
+    };
+    GsonBuilder gsonBuilder = new GsonBuilder();
+    for (Type type : types) {
+      IllegalArgumentException e = assertThrows(IllegalArgumentException.class,
+              () -> gsonBuilder.registerTypeAdapter(type, NULL_TYPE_ADAPTER));
+      assertThat(e).hasMessageThat().isEqualTo(ERROR_MESSAGE + type);
+    }
+  }
+
+
+  @Test
+  public void testRegisterTypeHierarchyAdapterJsonElements() {
+    final String ERROR_MESSAGE = "Cannot override built-in adapter for ";
+    Type[] types = {
+            JsonElement.class,
+            JsonArray.class,
+    };
+    GsonBuilder gsonBuilder = new GsonBuilder();
+    for (Type type : types) {
+      IllegalArgumentException e = assertThrows(IllegalArgumentException.class,
+              () -> gsonBuilder.registerTypeHierarchyAdapter((Class<?>) type, NULL_TYPE_ADAPTER));

```

Comment:
As mentioned in https://github.com/google/gson/pull/2479#discussion_r1306483982, you can avoid the `Class<?>` cast here if you want (and also the indentation seems to be incorrect):
```suggestion
    Class<?>[] types = {
        JsonElement.class,
        JsonArray.class,
    };
    GsonBuilder gsonBuilder = new GsonBuilder();
    for (Class<?> type : types) {
      IllegalArgumentException e = assertThrows(IllegalArgumentException.class,
          () -> gsonBuilder.registerTypeHierarchyAdapter(type, NULL_TYPE_ADAPTER));

```